src/crypto/hashing.rs

```diff
@@ -1 +1,239 @@
 
+use data_encoding::HEXUPPER;
+use sha2::{Sha256, Digest};
+use std::convert::TryInto;
+use std::ops;
+use std::fmt;
+
+// Definition of Hash struct
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Hash([u8; 32]);
+
+impl Hash {
+    // Method that takes a byte array and return hash
+    pub fn from_bytes(msg: &[u8]) -> Hash {
+        let mut hasher = Sha256::new();
+        hasher.update(msg);
+        let result = hasher.finalize();
+        Hash(result.as_slice().try_into().unwrap())
+    }
+
+    // Method that takes two Hash and return the Hash of their concatenation
+    pub fn concat(lhs: &Hash, rhs: &Hash) -> Hash {
+        let mut result = Vec::new();
+        result.extend_from_slice(&lhs.0);
+        result.extend_from_slice(&rhs.0);
+        let result_slice: &[u8] = &result;
+        Hash::from_bytes(&result_slice)
+    }
+}
+
+impl fmt::Display for Hash {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", HEXUPPER.encode(&self.0))
+    }
+}
+
+// Basically Hash::concat
+impl ops::Add for Hash {
+    type Output = Self;
+    fn add(self, rhs: Self) -> Self {
+        Hash::concat(&self, &rhs)
+    }    
+}
+
+// End of Hash 
+
+// Definition of Hashable
+pub trait Hashable {
+    fn hash(&self) -> Hash;
+}
+// End of hashable
+
+// Definition of Merkle Tree Node
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MerkleNode {
+    value: Hash,
+    left: i32,
+    right: i32,
+    size: usize,
+}
+
+impl MerkleNode {
+    pub fn merge(tree: &Vec::<MerkleNode>, left: usize, right: usize) -> MerkleNode {
+        let size = tree[left].size + tree[right].size;
+        let value = tree[left].value + tree[right].value + size.hash();
+        MerkleNode {
+            value,
+            left: left as i32,
+            right: right as i32,
+            size,
+        }
+    }
+}
+// End of Merkle Tree Node
+
+
+// Definition of Merkle Tree
+#[derive(Debug, Clone, PartialEq, Eq)]
```

Comment:
Just deriving Eq here is inefficient… maybe a custom implementation that just compares hash and size?